cachelib/allocator/CacheAllocator-inl.h

```diff
@@ -3509,69 +3325,49 @@ void CacheAllocator<CacheTrait>::evictForSlabRelease(
     if (item.isChainedItem()) {
       auto& expectedParent = item.asChainedItem().getParentItem(compressor_);
 
-      if (getNumTiers() == 1) {
-        // TODO: unify this with multi-tier implementation
-        // right now, taking a chained item lock here would lead to deadlock
-        const std::string parentKey = expectedParent.getKey().str();
-        auto l = chainedItemLocks_.lockExclusive(parentKey);
-
-        // check if the child is still in mmContainer and the expected parent is
-        // valid under the chained item lock.
-        if (expectedParent.getKey() != parentKey || !item.isInMMContainer() ||
-            item.isOnlyMoving() ||
-            &expectedParent != &item.asChainedItem().getParentItem(compressor_) ||
-            !expectedParent.isAccessible() || !expectedParent.hasChainedItem()) {
-          continue;
-        }
-
-        // search if the child is present in the chain
-        {
-          auto parentHandle = findInternal(parentKey);
-          if (!parentHandle || parentHandle != &expectedParent) {
-            continue;
-          }
-
-          ChainedItem* head = nullptr;
-          { // scope for the handle
-            auto headHandle = findChainedItem(expectedParent);
-            head = headHandle ? &headHandle->asChainedItem() : nullptr;
-          }
-
-          bool found = false;
-          while (head) {
-            if (head == &item) {
-              found = true;
-              break;
-            }
-            head = head->getNext(compressor_);
-          }
+      const std::string parentKey = expectedParent.getKey().str();
+      auto l = chainedItemLocks_.tryLockExclusive(parentKey);
 
-          if (!found) {
-            continue;
-          }
-        }
+      if (!l) {
+        continue;
+      } else if (expectedParent.getKey() != parentKey) {
+        continue;
+      } else if (&expectedParent != &item.asChainedItem().getParentItem(compressor_)) {
+        continue;
+      } else if (!expectedParent.isAccessible()) {
+        continue;
+      } else if (!expectedParent.hasChainedItem()) {
+        continue;
       }
 
       evicted = &expectedParent;
       token = createPutToken(*evicted);
       if (evicted->markForEviction()) {
+        XDCHECK_EQ(&expectedParent,&item.asChainedItem().getParentItem(compressor_));
         // unmark the child so it will be freed
-        item.unmarkMoving();
+        // TODO entire chain just gets evicted since moveForSlabRelease
+        // returns false
+        XDCHECK(!item.isMoving());
```

Comment:
Why this assert? Can't item still be moving in this case? E.g. dure to markForSlabRelease